web/ui/react-app/src/pages/graph/PanelList.tsx

```diff
@@ -1,219 +1,178 @@
-              <Alert color="danger">
-                <strong>Warning:</strong> Error fetching metrics list: {fetchMetricsError}
-              </Alert>
-            )}
-          </Col>
-        </Row>
-        {panels.map(({ id, options }) => (
-          <Panel
-            onExecuteQuery={this.handleExecuteQuery}
-            key={id}
-            options={options}
-            onOptionsChanged={opts => this.handleOptionsChanged(id, opts)}
-            useLocalTime={this.state.useLocalTime}
-            removePanel={() => this.removePanel(id)}
-            metricNames={metricNames}
-            pastQueries={pastQueries}
-            pathPrefix={pathPrefix}
-          />
-        ))}
-        <Button color="primary" className="add-panel-btn" onClick={this.addPanel}>
-          Add Panel
-        </Button>
-      </>
-    );
-  }
-}
+  return (
+    <>
+      {panels.map(({ id, options }) => (
+        <Panel
+          onExecuteQuery={handleExecuteQuery}
+          key={id}
+          options={options}
+          onOptionsChanged={opts =>
+            callAll(setPanels, updateURL)(panels.map(p => (id === p.id ? { ...p, options: opts } : p)))
+          }
+          removePanel={() =>
+            callAll(setPanels, updateURL)(
+              panels.reduce<PanelMeta[]>(
+                (acc, panel) => (panel.id !== id ? [...acc, { ...panel, key: `${acc.length}` }] : acc),
+                []
+              )
+            )
+          }
+          useLocalTime={useLocalTime}
+          metricNames={metrics}
+          pastQueries={queryHistoryEnabled ? historyItems : []}
+          pathPrefix={pathPrefix}
+        />
+      ))}
+      <Button color="primary" onClick={addPanel}>
+        Add Panel
+      </Button>
+    </>
+  );
+};
+
+const opts: RequestInit = {
+  cache: 'no-store',
+};
+
+const PanelList: FC<RouteComponentProps & PathPrefixProps> = ({ pathPrefix = '' }) => {
+  const memoizedFetch = useCallback(useFetch, [pathPrefix]);
+  const [delta, setDelta] = useState(0);
+  const [panels, setPanels] = useState<PanelMeta[]>([]);
+  const [useLocalTime, setUseLocalTime] = useLocalStorage('use-local-time', false);
+  const [enableQueryHistory, setEnableQueryHistory] = useLocalStorage('enable-query-history', false);
+
+  const { response: metricsResponse, error: metricsError } = memoizedFetch<string[]>(
+    `${pathPrefix}/api/v1/label/__name__/values`,
+    opts
+  );
+
+  const browserTime = new Date().getTime() / 1000;
+  const { response: timeResponse, error: timeError } = memoizedFetch<{ result: number[] }>(
+    `${pathPrefix}/api/v1/query?query=time()`,
+    opts
+  );
+
+  useEffect(() => {
+    setPanels(decodePanelOptionsFromQueryString(window.location.search));
+  }, []);
+
+  useEffect(() => {
+    if (timeResponse.data) {
+      const serverTime = timeResponse.data.result[0];
+      setDelta(Math.abs(browserTime - serverTime));
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
Same here.

Reply:
I should find a way to disable those warnings officially(in the project). It's annoying because react doesn't leave you to decide what should be the dependencies list. For example, here I don't want to include browserTime since its changed on every re-render but react claims that this should be part of it even it automatically include it on save

Reply:
Or maybe it's not a good idea to disable it globally as per https://github.com/facebook/create-react-app/issues/6880